cyw43439: add support for CYW43439 wifi chip

irina-nita opened the pull request:
### Pull Request Overview

This pull request adds:
- `chips/cyw43439`: support for the PIO half-duplex SPI interface for the RP2040 chip used to communicate with the CYW43439 chip available on the Raspberry Pi Pico W board
- support in `boards/raspberry_pi_pico` for enabling the PIO configuration and Ethernet Tap capsule
- `hil/wifi_cyw43`: minimal WiFi HIL for peripherals that interact with the CYW43439 or other similar Cypress chips. The HIL contains four traits that provide interfaces for initializing the device and getting the MAC address, scanning networks and configuring station/access point mode.

### Testing Strategy

- [x] WiFi HIL implementation tested by scanning/connecting to/creating networks
- [x] Ethernet HIL tested with the `libtock-c` lwIP example

### TODO or Help Wanted

This PR needs to add an abstract interface for communicating with the CYW4343x chips. Besides the peripheral used for reading/writing to the chip, the process should be the same

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

alexandruradovici commented on chips/rp2040/src/pio.rs:
```diff
@@ -1786,6 +1829,45 @@ mod examples {
             sm.set_enabled(true);
         }
 
+        pub fn cyw43_spi_program_init(
```
Is this just an example?

alexandruradovici commented on kernel/src/hil/wifi_cyw43.rs:
```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OxidOS Automotive 2025.
+
+//! Interfaces for CYW43439 WiFi devices
```
Can we document here how these traits relate to each other and how they should be used?

alexandruradovici reviewed (changes requested):
Looks really nice, just a few ideas.

irina-nita commented on boards/raspberry_pi_pico/src/main.rs:
```diff
@@ -50,7 +53,7 @@ mod flash_bootloader;
 #[no_mangle]
 #[link_section = ".stack_buffer"]
-static mut STACK_MEMORY: [u8; 0x1500] = [0; 0x1500];
+static mut STACK_MEMORY: [u8; 0x6000] = [0; 0x6000];
```
I missed to modify this after adding a static buffer for packet construction. Fixed it now

irina-nita commented on boards/cyw43-firmware/43439A0.bin:
Created a `firmware` directory in root

bradjc commented on firmware/cyw43439/43439A0.bin:
I think it would be fine to put this in the board's directory.

Adding a new top-level directory is probably not something we want to do. Putting it in `chips/` is potentially an option, but I think would require some discussion and a plan for what types of binary blobs are acceptable and under what situations.

Another option is to embed the bytes in a rust file.

brghena replied:
We discussed this on the Network working group call and we're considering an external crate that the CYW* WiFi capsule would rely on which would contain the binary blobs. Notably, they're licensed differently than the rest of the Tock kernel and there are some concerns there.

bradjc replied:
That would probably still be easiest to include from the board, then.

bradjc commented on kernel/src/hil/wifi_cyw43.rs:
This seems general enough to just be called `wifi.rs`.

alexandruradovici replied:
Event though we discussed this during type networking call, I kind of agree with @bradjc that this seems general enough. I do not see how other devices might need anything else here.

alevy replied:
This is now totally superfluous since the cyw43439 is in a capsule (which is great). At least I can build everything no problem by just removing this, so I suggest we remove it.

bradjc commented on chips/rp2040/src/cyw43439/packets.rs:
```diff
@@ -0,0 +1,712 @@
+}
+
+#[macro_export]
+macro_rules! impl_bytes {
+    ($t:ident) => {
+        impl $t {
+            /// Bytes consumed by this type.
+            pub const SIZE: usize = core::mem::size_of::<Self>();
+
+            /// Create from byte array.
+            #[allow(unused)]
+            pub fn from_bytes(bytes: &[u8; Self::SIZE]) -> &Self {
+                let alignment = core::mem::align_of::<Self>();
+                assert_eq!(
+                    bytes.as_ptr().align_offset(alignment),
+                    0,
+                    "{} is not aligned",
+                    core::any::type_name::<Self>()
+                );
+                unsafe { core::mem::transmute(bytes) }
+            }
+
+            /// Create from mutable byte array.
+            #[allow(unused)]
+            pub fn from_bytes_mut(bytes: &mut [u8; Self::SIZE]) -> &mut Self {
+                let alignment = core::mem::align_of::<Self>();
+                assert_eq!(
+                    bytes.as_ptr().align_offset(alignment),
+                    0,
+                    "{} is not aligned",
+                    core::any::type_name::<Self>()
+                );
+
+                unsafe { core::mem::transmute(bytes) }
+            }
+        }
+    };
+}
```
I'm worried about having these unsafes here in general, and based on learning how macros work in Rust having `unsafe` in a macro is even more scary. It's not clear to me what `impl_bytes` does or how necessary it is versus using a safe approach.

brghena replied:
This file might also be copied directly from Embassy, if I recall what we discussed on the Network working group call. Maybe we should be re-implementing these for Tock.

bradjc commented on chips/rp2040/src/cyw43439/utils.rs:
```diff
@@ -0,0 +1,185 @@
+}
+}
+
+pub(super) fn as_bytes<T: Sized>(data: &T) -> &[u8] {
+    unsafe {
+        core::slice::from_raw_parts(
+            core::ptr::from_ref::<T>(data) as *const u8,
+            core::mem::size_of::<T>(),
+        )
+    }
+}
+
+pub(super) fn slice8_mut(x: &mut [u32]) -> &mut [u8] {
+    let len = x.len() * 4;
+    unsafe { slice::from_raw_parts_mut(x.as_mut_ptr() as _, len) }
+}
+
```
With these unsafes I don't know how necessary they are, but it would be good to add a `# Safety` comment that explains why they are safe.

alexandruradovici commented on kernel/src/hil/wifi_cyw43.rs:
```diff
@@ -0,0 +1,184 @@
+    /// ## Arguments
+    ///
+    /// - `ssid`: SSID of found network
+    fn scanned_network(&self, ssid: Ssid);
```
```suggestion
    fn scanned_network(&self, ssid: Ssid, encryption: Option<Encryption>, signal_strength: u8);
```

```rust
#[non_exhaustive]
enum Encryption {
   Open,
   Wep,
   Wpa1,
   Wpa2,
   Wpa3,
}
```

Just to make it more generic.

irina-nita commented:
Hi, sorry it took this long 😅. So far I have managed to:
- move the logic from `chips` to `capsules/cyw4343` and get rid of the unsafe packet parsing
- add DMA support for the RP2040 so the transfers aren't blocking anymore
- add a `wifi` capsule, still some TODOs there, but it's working. I've tested it with a [libtock-c app](https://gist.github.com/irina-nita/66bb825054c69b2e7232cbd3b4f8fc14)
- removed ethernet support for now from the board's main (still not sure how it would work alongside wifi)

bradjc commented on capsules/cyw4343/src/lib.rs:
```diff
@@ -0,0 +1,1219 @@
+
+//! Driver for Infineon's CYW4343x WiFi chips
+
+#![no_std]
```
```suggestion
#![forbid(unsafe_code)]
#![no_std]
```
?

bradjc reviewed (changes requested):
I don't think including the binary blob external crate in capsules is consistent with our external dependencies guide. Particularly:

https://github.com/tock/tock/blob/4ad42c94bb66b8204f5e14e8f0b986521d8a6f9e/doc/ExternalDependencies.md?plain=1#L117-L122

I think it does meet the criteria for a board-level external dependency. I don't remember if we came to some other conclusion on a call, however. If so, I think we need a fairly clear explanation of why.

Other changes needed:

- Output of cargo tree for the dependency
- Squash commit history to remove firmware.

brghena commented on boards/raspberry_pi_pico/src/main.rs:
```diff
@@ -77,6 +80,8 @@ type TemperatureRp2040Sensor = components::temperature_rp2040::TemperatureRp2040
 >;
 type TemperatureDriver = components::temperature::TemperatureComponentType<TemperatureRp2040Sensor>;
 
+const WIFI: bool = option_env!("WIFI").is_some();
```
Can we just make WiFi enabled by default for the raspberry_pi_pico board? It's not clear that anyone would want to disable it if it works.

Generally, I'm pretty against environment variables like this. They're hidden state that is hard to debug for new users. Advanced users can always modify the board file themselves if they don't want WiFi.

brghena commented on capsules/cyw4343/src/sdpcm.rs:
```diff
@@ -0,0 +1,601 @@
+
+use capsules_extra::wifi;
+
+macro_rules! parse {
```
It would be great to have a big comment here about what this does and why. It's used heavily in the file, but also pretty hard to read.

brghena commented on capsules/cyw4343/README.md:
```diff
@@ -0,0 +1,6 @@
+=======================
+
+This crate adds support for the CYW4343x WiFi chips.
+
```
Could you include a description of each file in this folder and what its purpose is?

irina-nita replied:
I added top-level comments for the modules since I moved them to `capsules/extra`

brghena commented on boards/raspberry_pi_pico/src/main.rs:
```diff
@@ -439,10 +459,12 @@ pub unsafe fn start() -> (
+            LedHigh<'static, RPGpioPin<'static>>,
+            LedHigh::new(peripherals.pins.get_pin(RPGpio::GPIO25))
+        ))
+    });
```
We might be able to use a mechanism like Configuration Boards so that the two Pico boards share 90%+ of their code and only differ in ways that they are actually different. https://github.com/tock/tock/tree/master/boards/configurations

I'm not 100% sure how to do that, but that's the goal. Maybe @lschuermann can provide additional details.

lschuermann replied:
Right, the basic idea is to share the bulk of the board file's code between the two boards, by converting one into a shared library+binary crate like we've done with the nRF52840DK: https://github.com/tock/tock/tree/master/boards/nordic/nrf52840dk/src

Here, the `lib.rs` file is part of a "library crate" that can be imported from within other boards, like so: https://github.com/tock/tock/blob/6b9da6489887c9d3b3061d5a94af87221e163e36/boards/tutorials/nrf52840dk-hotp-tutorial/Cargo.toml#L20

Then, you have two boards defined in two `main.rs` files; one in the combined library+binary crate (which could be the Pico non-W): https://github.com/tock/tock/blob/master/boards/nordic/nrf52840dk/src/main.rs

And another one (which would be the Pico W) in the crate importing the base board definition: https://github.com/tock/tock/blob/master/boards/tutorials/nrf52840dk-hotp-tutorial/src/main.rs

bradjc reviewed (dismissed):
I agree with Branden's comments but want to clear my red x.

irina-nita commented:
(sorry for the commits mess, it's 3AM in Romania 😅)
I fixed the passphrase representation in the Wi-Fi capsule because it was the CYW4343 IOCTL size, not the actual standard size. I removed the SCAN_DONE upcall. Now the scanning is done when the SCAN_RES upcall is sent with 0 as first argument insted of the length (I also enforced that the SSID length is non-zero with a `NonZeroU8`).

bradjc commented on boards/components/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ capsules-core = { path = "../../capsules/core" }
 capsules-extra = { path = "../../capsules/extra" }
 capsules-system = { path = "../../capsules/system" }
 segger = { path = "../../chips/segger" }
+rp2040 = { path = "../../chips/rp2040" }
```
```suggestion
```

bradjc commented on boards/raspberry_pi_pico_w/src/main.rs:
```diff
@@ -0,0 +1,209 @@
+pub struct RaspberryPiPicoW {
+    base: raspberry_pi_pico::Platform,
+    wifi: &'static capsules_extra::wifi::WifiDriver<'static, CYW4343xDriver>,
+}
```
```suggestion
type WifiDriver = capsules_extra::wifi::WifiDriver<'static, CYW4343xHw>

/// Supported drivers by the platform
pub struct RaspberryPiPicoW {
    base: raspberry_pi_pico::Platform,
    wifi: &'static WifiDriver,
}
```

bradjc commented on boards/raspberry_pi_pico_w/src/main.rs:
```diff
@@ -0,0 +1,209 @@
+    RPGpioPin<'static>,
+    VirtualMuxAlarm<'static, rp2040::timer::RPTimer<'static>>,
+    CYW4343xSpiBus,
+>;
```
```suggestion
type CYW4343xHw = capsules_extra::cyw4343::CYW4343x<
    'static,
    RPGpioPin<'static>,
    VirtualMuxAlarm<'static, rp2040::timer::RPTimer<'static>>,
    CYW4343xSpiBus,
>;
```

Ending in `Driver` is for syscall drivers.

brghena commented on boards/raspberry_pi_pico/layout.ld:
```diff
@@ -11,8 +11,8 @@ MEMORY
-  rom (rx)  : ORIGIN = 0x10000000, LENGTH = 256K
-  prog (rx) : ORIGIN = 0x10040000, LENGTH = 256K
+  rom (rx)  : ORIGIN = 0x10000000, LENGTH = 576K
+  prog (rx) : ORIGIN = 0x10090000, LENGTH = 256K
```
Was this wrong before?

brghena commented on capsules/extra/src/cyw4343/mod.rs:
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OxidOS Automotive 2025.
+
+//! CYW4343x driver
```
Could you put a small README.md in the `capsules/extra/src/cyw4343/` directory that explains what the device is, links to any references you have, and explains (very briefly) what each of these files are?

For instance, I don't know what "sdpcm" might stand for. And I'm not sure how these files are organized or connect to each other.

This doesn't have to be a huge README, as even a little documentation would help quite a bit.

irina-nita replied:
Added a small README. Regarding what SDPCM exactly stands for, I couldn't find any official documentation; my hunch was that it stood for something like "SDIO Packet Multiplexor"

lschuermann commented on boards/raspberry_pi_pico_w/src/main.rs:
```diff
@@ -0,0 +1,213 @@
+    let pwr = peripherals.pins.get_pin(RPGpio::GPIO23);
+    pwr.make_output();
+
+    let bus =
```
Nit, but can we rename this to something more meaningful than `bus`? Perhaps `cyw4343_spi_bus`?

lschuermann commented on boards/raspberry_pi_pico_w/layout.ld:
As commented on the other linker script by @brghena already, this should probably be a separate copy with the appropriate changes required for WiFi support.

lschuermann commented on boards/raspberry_pi_pico_w/README.md:
Can we add some getting started documentation here for getting this to join a WiFi network, or is this something best left to a userspace repository and/or follow-up PR?

lschuermann commented on capsules/extra/src/cyw4343/utils.rs:
Not so much utilities than constants, no? Perhaps rename to `constants.rs`?

lschuermann commented on chips/rp2040/src/dma.rs:
```diff
@@ -0,0 +1,707 @@
+                    .map(|client| client.transfer_done());
+            }
+        }
+    }
```
Or, even better, [`lowest_one`](https://doc.rust-lang.org/stable/std/primitive.usize.html#method.lowest_one). This is nightly-only at this point, but you could vendor the implementation.

lschuermann commented on chips/rp2040/src/pio.rs:
```diff
@@ -943,6 +954,24 @@ impl StateMachine {
+    }
+
+    /// Address of the TX FIFO
+    pub fn txf_addr(&self, pio: PIONumber) -> u32 {
```
```suggestion
    pub fn tx_fifo_addr(&self, pio: PIONumber) -> u32 {
```

irina-nita commented on capsules/extra/src/wifi/device.rs:
```diff
@@ -0,0 +1,134 @@
+//! Hardware-independent interface for WiFi devices.
+//!
+//! This interface provides high-level functionalities: scanning for networks, joining a network,
+//! configure as access point, get MAC address.
```
As you've mentioned in your other comment regarding the capsule name, `IEEE80211Control` device/driver sounds better (similar to how the [`cfg80211` subsystem](https://www.kernel.org/doc/html/v4.12/driver-api/80211/cfg80211.html#) is named in the Linux kernel; the `Device` API I think should provide a similar interface to `cfg80211_ops`).
Should the renaming be added in this PR?
Also, the interface is very limited right now :( . In the future I'll make another PR improving this with power management and other actions.

irina-nita commented on capsules/extra/src/wifi/device.rs:
```diff
@@ -0,0 +1,134 @@
+        Wpa2 = 2,
+        Wpa2Wpa3 = 3,
+        Wpa3 = 4,
+    }
```
I modified the enum variants. I know this is still the very bare minimum and I was thinking about extending the interface, but it would also require some changes in the `driver` module and the `cyw4343` driver also. Should this be done in the current PR?

irina-nita commented on capsules/extra/src/cyw4343/bus/common.rs:
```diff
@@ -0,0 +1,107 @@
+
+#[derive(Clone, Copy)]
+pub(crate) enum BackplaneTask {
+    // TODO(?): Maybe move the function pointer to an enum
```
Nope

irina-nita commented on capsules/extra/src/cyw4343/bus/mod.rs:
```diff
@@ -0,0 +1,117 @@
+//!
+//! F2 is used for reading/writing WLAN packets. The bus implementation should only handle
+//! sending data as F2 packets and it should expect that the protocol headers have been added by
+//! the upper layer.
```
I updated the comment specifying that I meant the SDPCM protocol headers

irina-nita commented on capsules/extra/src/cyw4343/sdpcm.rs:
````diff
@@ -0,0 +1,626 @@
+///
+/// ## Example
+///
+/// ```rust,ignore
````
Modified it but reverted back to this because it fails due to the macro being in a private module 😅
